data_structures/monotonic_stack.cpp

```diff
@@ -0,0 +1,109 @@
+/**
+ * @file
+ * @brief Implementation of Monotonic Stack
+ * @details
+ * This is an implementation of a monotonic stack which maintains the stack in increasing order.
+ * Elements are pushed such that the stack remains monotonic.
+ * If an element is pushed that is smaller than the top element, the top elements are popped until the correct position is found.
+ * @see https://en.wikipedia.org/wiki/Monotonic_Stack
+ * @author [Srijan Baniyal](https://github.com/Srijan-Baniyal)
+ */
+
+#include <iostream>        /// For IO operations
+#include <vector>         /// For std::vector
+#include <stdexcept>      /// For std::out_of_range
+
+/**
+ * @namespace data_structures
+ * @brief A namespace for various data structure implementations
+ */
+namespace data_structures {
+
+/**
+ * @class MonotonicStack
+ * @brief A class representing a monotonic stack which maintains elements in increasing order.
+ * @tparam T the type of elements in the stack (must support comparison operators)
+ */
+template <typename T>
+class MonotonicStack {
+ private:
+    std::vector<T> stack;   ///< Vector to store stack elements
```

Comment:
this stack is heap allocated should it not be stack allocated? i.e use a std::array with a max size?

Reply:
ok I'll change it.